core/stats.go

```diff
@@ -57,24 +69,60 @@ func (ds *dockerService) ListContainerStats(
 		return nil, err
 	}
 
-	var mtx sync.Mutex
-	var wg sync.WaitGroup
-	var stats = make([]*runtimeapi.ContainerStats, 0, len(listResp.Containers))
-	for _, container := range listResp.Containers {
-		container := container
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			if containerStats, err := ds.getContainerStats(container.Id); err == nil && containerStats != nil {
-				mtx.Lock()
-				stats = append(stats, containerStats)
-				mtx.Unlock()
-			} else if err != nil {
-				logrus.Error(err, " Failed to get stats from container "+container.Id)
+	g, ctx := errgroup.WithContext(ctx)
+	containerIDCh := make(chan string)
+
+	g.Go(func() error {
+		defer close(containerIDCh)
+		for _, container := range listResp.Containers {
+			select {
+			case containerIDCh <- container.Id:
+			case <-ctx.Done():
```

Comment:
`case <-ctx.Done()` here is used for return early after the context is cancelled. It means a worker failed or the client has cancelled the (parent) context. 

Successfully iterating all the containers here just means we finish sending the jobs to the job queue `containerIDCh`, the jobs are not finished yet, thus it will not receive anything from `<-ctx.Done()` unless an error happens.